feat: swaps new asset picker

bfullam opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

Replaces the legacy swap asset picker interfaces with a new, unified asset selector. The new selector is full screen and utilized the new /popular and /search endpoints of the bridge API for significantly faster loading times.

Design: https://www.figma.com/design/1F3yNWYLOVPFpTPeJugH20/SWAP?node-id=9898-17382&t=7aukGCuzC3p7zFV0-0

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and accurately)
-->

CHANGELOG entry: Added new swaps asset picker

## **Related issues**

Fixes:

## **Manual testing steps**

```gherkin
Feature: new asset picker

  Scenario: user wants to select a swap source or destination asset
    Given they are on the swap page

    When user clicks asset icon
    Then full screen asset picker opens and allows asset selection
```

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**


https://github.com/user-attachments/assets/c25bc108-b237-4bab-869c-8a3c613972fd

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.




<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Consolidates Bridge token selection and cleans up deprecated screens.
> 
> - Route changes: `BridgeView` now navigates to `Routes.BRIDGE.TOKEN_SELECTOR` with `{ type: 'source' | 'dest' }`; updated tests to assert new route/params
> - Removes legacy `BridgeDestNetworkSelector`, `BridgeDestTokenSelector`, `BridgeDestNetworksBar` and their tests/snapshots
> - Adds `isSelectingToken` to view state to prevent quote-expired modal while selecting; updates related mocks
> - Simplifies fee disclaimer rendering to always show `bridge.no_mm_fee_disclaimer` when no fee (removes `noFeeDestAssets` check)
> - Component library: `KeyValueRowTooltip` accepts `bottomPadding`; `KeyValueRowLabel` passes it to `openTooltipModal`
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 4d71c7f9036f4abe94e2cb11cddf29ee40e03c5b. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

bfullam committed:
feat: add modal scaffolding and network pills

bfullam committed:
feat: added searchbar

bfullam committed:
feat: add token flatlist

bfullam committed:
chore: navigate to new token selector

bfullam committed:
feat: add asset list

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
chore: prevent re-renders on token balance change

bfullam committed:
feat: add search

bfullam committed:
feat: add search pagination

bfullam committed:
chore: extract styles

bfullam committed:
chore: extract populat token fetching to hook

bfullam committed:
chore: extract token type conversion to util function

bfullam committed:
chore: prettier

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
chore: extract token search to hook

bfullam committed:
chore: remove hasSearchedOnce ref

bfullam committed:
chore: cleanup and reorg

bfullam committed:
chore: use API data for noFee badge

bfullam committed:
chore: fix metrics

bfullam committed:
feat: set source and dest token on selection

bfullam committed:
chore: streamline "no fee disclaimer" logic

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
feat: add request caching

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
fix: prevent race condition on quick network switch

bfullam committed:
chore: remove network icon from network pills

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
refactor: use screen view instead of modal

bfullam committed:
chore: use bridge api constants instead of hardcoded values

bfullam committed:
chore: perf improvements

bfullam committed:
fix: fix lazy load on search

bfullam committed:
refactor: transition from excludeAssetIds to includeAssets for token endpoints

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
chore: use shortened network names

bfullam committed:
chore: transition to tailwind

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
chore: lint

bfullam committed:
test: update tests

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
chore: streamline useBalancesByAssetId

bfullam committed:
fix: fix "all networks" search functionality

bfullam committed:
chore: filter out non-tradable tokens

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
chore: fix convertAPITokensToBridgeTokens function

bfullam committed:
chore: always open to "all" networks

bfullam committed:
chore: only pre-set chain on dest token picker and auto scroll to that pill on load

bfullam committed:
fix: adjust ethereum network pill auto scroll

bfullam committed:
fix: improve lazyload after network change

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
chore: update network pill styling

bfullam committed:
feat: keep search term on network change

bfullam committed:
chore: use min 3 character search length

bfullam committed:
fix: fix isSelected logic

bfullam committed:
feat: switch tokens when selecting already selected token

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
feat: auto load second page of search results

bfullam committed:
chore: streamlining refactor

bfullam committed:
fix: fix lazyloading after network switch

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
chore: prettier

bfullam committed:
test: add unit tests

bfullam committed:
chore: fix test

bfullam committed:
test: coverage

bfullam committed:
chore: prettier

bfullam committed:
chore: sonarqube reliability fixes

bfullam committed:
test: streamline tests

bfullam committed:
feat: use new asset info modal

bfullam committed:
chore: normalize assetIds for EVM to prevent asset duplication

bfullam committed:
test: update tests

bfullam committed:
test: fix test

bfullam committed:
chore: sonarqube reliability

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
test: update bridge e2e mocks

bfullam committed:
test: update view test

bfullam committed:
test: add popular tokens request mock

bfullam committed:
chore: lint

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
test: update e2e

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
test: update test

bfullam committed:
test: add e2e mocks

cortisiko reviewed (dismissed):
approving from an e2e test perspective

GeorgeGkas commented on app/components/UI/Bridge/components/BridgeTokenSelector/BridgeTokenSelector.tsx:
```diff
@@ -0,0 +1,443 @@
+  // Initialize selectedChainId with the chain id of the selected token
+  const [selectedChainId, setSelectedChainId] = useState<
+    CaipChainId | undefined
+  >(
```
no need to explicitly set the type, it can be inferred from the expression.

bfullam replied:
addressed

GeorgeGkas commented on app/components/UI/Bridge/components/BridgeTokenSelector/BridgeTokenSelector.tsx:
```diff
@@ -0,0 +1,443 @@
+
+  // Chain IDs to fetch tokens for
+  const chainIdsToFetch = useMemo(() => {
+    // @ts-expect-error chainRanking is not yet in the type definition
```
can we update the type definition of feature flags to include the missing property rather than using expect error? I am afraid we forget to fix it later. Even if this is not the case, if the changes are not too complicated, I say lets fix it gracefully.

GeorgeGkas commented on app/components/UI/Bridge/components/BridgeTokenSelector/BridgeTokenSelector.tsx:
```diff
@@ -0,0 +1,443 @@
+    // If "All" is selected, use all chains from chainRanking
+    // @ts-expect-error chainRanking is not yet in the type definition
+    return bridgeFeatureFlags.chainRanking.map(
+      (chain: { chainId: CaipChainId }) => chain.chainId,
```
...this way the callback param won't be `any` and won't need to explicitly define its type.

GeorgeGkas commented on app/components/UI/Bridge/components/BridgeTokenSelector/BridgeTokenSelector.tsx:
```diff
@@ -0,0 +1,443 @@
+  // Create includeAssets array from tokens with balance to be sent to API
+  // Stringified to avoid triggering the useEffect when only balances change
+  const includeAssets = useMemo(() => {
+    const assets: IncludeAsset[] = filteredTokensWithBalance.reduce<
```
no need to define the type for `assets` as you pass it as template parameter in `reduce`.

GeorgeGkas commented on app/components/UI/Bridge/components/BridgeTokenSelector/BridgeTokenSelector.tsx:
```diff
@@ -0,0 +1,443 @@
+    (event: NativeSyntheticEvent<NativeScrollEvent>) => {
+      const { layoutMeasurement, contentOffset, contentSize } =
+        event.nativeEvent;
+      const paddingToBottom = 300; // Distance from bottom to trigger load
```
we can probably find a more descriptive name for it like `LOAD_MORE_DISTANCE_THRESHOLD` and move it to the top of the file like the other constants.

GeorgeGkas commented on app/components/UI/Bridge/components/BridgeTokenSelector/NetworkPills.tsx:
```diff
@@ -0,0 +1,147 @@
+    ? (`0x${parseInt(chainId.split(':')[1]).toString(16)}` as Hex)
+    : chainId;
+
+  return (
```
to be honest, I do not like the semantics of this expression, it relies heavily on nullish coalescing assignment, making this difficult to understand. I would prefer if we could split it into multipe statements or just add two lines of comments explaining the logic.

GeorgeGkas commented on app/components/UI/Bridge/components/BridgeTokenSelector/NetworkPills.tsx:
```diff
@@ -0,0 +1,147 @@
+    // Only scroll if the selected network is beyond the first 2 visible networks
+    // The first few networks are already visible, no need to scroll
+    if (selectedIndex > 1) {
+      const pillWidth = 90; // Average pill width including gap
```
we can follow the same pattern as with other constants and move this into top of the file.

GeorgeGkas commented on app/components/UI/Bridge/hooks/useBalancesByAssetId/index.ts:
```diff
@@ -0,0 +1,63 @@
+        // Normalize assetId because API returns assetId in lowercase for EVM chains
+        const normalizedAssetId = isNonEvmChainId(token.chainId)
+          ? assetId
+          : assetId?.toLowerCase();
```
assetId is always a string based on its type.

bfullam replied:
addressed

GeorgeGkas reviewed (changes requested):
🚀 🚀 🚀 🚀 🚀 🚀 🚀 

Most of my comments are about simplifying typings and making a few stylistic updates. Also, please take some time to review the addition of chainRanking to our types as using @ts-expect-error comments isn’t ideal for long-term maintainability.

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
fix: don't clear search results when already selected network pill is clicked

bfullam committed:
chore: spacing fixes

bfullam committed:
chore: remove deprecated components

bfullam committed:
test: update tests

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
fix: fix lazyloading after network change

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
chore: disable autocomplete on asset picker search bar

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
fix: fix lazyloading on search after network change

bfullam committed:
test: update tests

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
feat: add empty search state

bfullam committed:
test: fix tests

bfullam committed:
chore: stylistic changes

bfullam commented on app/components/UI/Bridge/components/BridgeTokenSelector/BridgeTokenSelector.tsx:
```diff
@@ -0,0 +1,443 @@
+      flatListHeight > 0
+    ) {
+      // Estimate item height (approximate height of TokenSelectorItem)
+      const ESTIMATED_ITEM_HEIGHT = 72;
```
addressed

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

metamaskbot committed:
[skip ci] Bump version number to 3375

GeorgeGkas committed:
fix: render fallback token icon in asset insights bottom sheet

bfullam committed:
fix: render fallback token icon in asset insights bottom sheet (#24307)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

Render fallback token icon in asset insights bottom sheet

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the
CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing
description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and
accurately)
-->

CHANGELOG entry: render fallback token icon in asset insights bottom
sheet

## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/SWAPS-3604

## **Manual testing steps**

```gherkin
Feature: my feature name

  Scenario: user [verb for user action]
    Given [describe expected initial app state]

    When user [verb for user action]
    Then [describe expected outcome]
```

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<img width="540" height="348" alt="Screenshot 2026-01-08 at 10 51 20 AM"
src="https://github.com/user-attachments/assets/b01ecffe-5575-47db-afc5-24816e9a71d0"
/>


<!-- [screenshots/recordings] -->

### **After**
<img width="540" height="348" alt="Screenshot 2026-01-08 at 10 49 39 AM"
src="https://github.com/user-attachments/assets/073ac73e-67af-4649-b6c7-7b0251887921"
/>

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Switches token avatar rendering to a component with built-in fallback
handling.
> 
> - Replaces `TokenIcon` with `AvatarToken` in `TokenInsightsSheet` to
render `imageSource` when available and fallback icon otherwise
> - Adds tests for avatar display with image, missing image, and empty
image in `TokenInsightsSheet.test.tsx`
> - Updates Jest snapshots to reflect new avatar structure and testIDs
> 
> <sup>Written by [Cursor
Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit
22696e21e33d04aa2dd2947da4edfd702414cfa1. This will update automatically
on new commits. Configure
[here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

bfullam committed:
chore: remove "view all tokens" button

bfullam committed:
chore: don't show refresh banner while on asset picker

bfullam committed:
chore: remove disabled networks from asset picker network pills

bfullam committed:
chore: prettier

bfullam committed:
feat: add clear search button

bfullam committed:
feat: add optional padding to tooltip modals

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

GeorgeGkas committed:
fix: select token title issues on android and ios

bfullam committed:
chore: navigate to asset detail page instead of token insights modal

GeorgeGkas committed:
test: update

GeorgeGkas committed:
Merge branch 'swaps-new-token-picker' into swaps-3676-title-not-centered

GeorgeGkas committed:
Merge branch 'main' into swaps-new-token-picker

GeorgeGkas committed:
Merge branch 'swaps-new-token-picker' into swaps-3676-title-not-centered

GeorgeGkas committed:
test: updating outdated test

bfullam committed:
fix: select token title issues on android and ios (#24436)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

- Set `headerTitleAlign` explicitly as Android does not default to
center.
- Fix an issue with overlapping titles in ios.

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the
CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing
description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and
accurately)
-->

CHANGELOG entry: fixes issues with select token title on android and ios

## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/SWAPS-3676

## **Manual testing steps**

```gherkin
Feature: my feature name

  Scenario: user [verb for user action]
    Given [describe expected initial app state]

    When user [verb for user action]
    Then [describe expected outcome]
```

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<img width="540" height="976" alt="Screenshot 2026-01-12 at 1 04 17 PM"
src="https://github.com/user-attachments/assets/b4509be0-5fc5-4ca0-9013-b9b05d0598b3"
/>


<!-- [screenshots/recordings] -->

### **After**
<img width="720" height="605" alt="image"
src="https://github.com/user-attachments/assets/076c946c-16b0-4793-aa74-9699ff275761"
/>

<img width="540" height="454" alt="Screenshot 2026-01-12 at 1 27 02 PM"
src="https://github.com/user-attachments/assets/8529d95a-96c4-49ed-a0ba-359a21fa53f2"
/>

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> - **Navbar/Navigation:** Centers `BridgeTokenSelector` title via
`getBridgeTokenSelectorNavbar` and enables headers for
`BridgeScreenStack` (`headerMode="screen"`, `headerShown: true`).
> - **Tooltips:** Adds `bottomPadding: 64` to tooltip modal navigations
in `QuoteDetailsCard` tests.
> - **Asset details nav:** `BridgeTokenSelector` info button now
navigates to `Asset` with token details (symbol, name, assetId, chainId,
decimals, image) instead of `RootModalFlow/TokenInsights`.
> - **Tests:** Updated expectations in `BridgeTokenSelector.test.tsx`
and `QuoteDetailsCard.test.tsx` to reflect the above behaviors.
> 
> <sup>Written by [Cursor
Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit
02e69609158f62a962f26a7ac55201cc5d2dd8a9. This will update automatically
on new commits. Configure
[here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

brianacnguyen reviewed (changes requested):
Can you add screen recordings and design links?

GeorgeGkas committed:
chore: Merge branch 'main' into swaps-new-token-picker

GeorgeGkas committed:
fix: unified asset image rendering on asset picker

bfullam committed:
fix: unified asset image rendering on asset picker (#24463)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

Fix an issue with native assets showing a fade in animation. 

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the
CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing
description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and
accurately)
-->

CHANGELOG entry: Fix an issue with native assets showing a fade in
animation.

## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/SWAPS-3602

## **Manual testing steps**

```gherkin
Feature: my feature name

  Scenario: user [verb for user action]
    Given [describe expected initial app state]

    When user [verb for user action]
    Then [describe expected outcome]
```

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Unifies token image rendering in `TokenSelectorItem` to use
`AvatarToken` for both native and ERC20 assets.
> 
> - Replaced `TokenIcon` branch with a single `AvatarToken` and set
`testID` to `network-logo-<symbol>` for native assets and
`token-logo-<symbol>` otherwise
> - Removed unused `nativeTokenIcon` styles and `TokenIcon` mocks;
updated test to reflect native token rendering without TokenIcon
> - No functional changes to balance/labels; visuals and tests now
consistent across assets
> 
> <sup>Written by [Cursor
Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit
29b8f092651437ad336aba503e5887febb269e68. This will update automatically
on new commits. Configure
[here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

GeorgeGkas committed:
fix: supply missing accountType field on token object

GeorgeGkas committed:
fix: set max display size for asset names in new swaps asset picker

GeorgeGkas committed:
fix: reduce the number of lines to two

bfullam committed:
fix: supply missing accountType field on token object (#24475)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

Add missing field to token images to allow asset labels to be rendered.

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the
CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing
description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and
accurately)
-->

CHANGELOG entry: Add missing field to token images to allow asset labels
to be rendered.

## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/SWAPS-3748

## **Manual testing steps**

```gherkin
Feature: my feature name

  Scenario: user [verb for user action]
    Given [describe expected initial app state]

    When user [verb for user action]
    Then [describe expected outcome]
```

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Ensures `accountType` is included and preserved across balance lookups
and token merging.
> 
> - Extend `BalanceData` with `accountType` and include it when building
`balancesByAssetId` in `useBalancesByAssetId`
> - Merge `accountType` from `balancesByAssetId` into tokens in
`useTokensWithBalances`
> - Add/expand tests in `useBalancesByAssetId/index.test.ts` and
`useTokensWithBalances.test.ts` to cover presence/absence of
`accountType` and preserve other optional fields
> 
> <sup>Written by [Cursor
Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit
e5c13065ff434103d6daf0a3baaa4f8b31cb9d2c. This will update automatically
on new commits. Configure
[here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

GeorgeGkas committed:
fix: set max display size for asset names in new swaps asset picker (#24486)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the
CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing
description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and
accurately)
-->

CHANGELOG entry: set max display size for asset names in new swaps asset
picker

## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/SWAPS-3754

## **Manual testing steps**

```gherkin
Feature: my feature name

  Scenario: user [verb for user action]
    Given [describe expected initial app state]

    When user [verb for user action]
    Then [describe expected outcome]
```

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<img width="333" height="720" alt="εικόνα"
src="https://github.com/user-attachments/assets/8662dd0b-fef8-4e1f-87be-26748d9363cb"
/>


<!-- [screenshots/recordings] -->

### **After**

<img width="719" height="306" alt="εικόνα"
src="https://github.com/user-attachments/assets/225b1069-7c10-4a8b-abcb-b692321d067c"
/>


<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Improves readability of long asset names in the bridge token selector.
> 
> - In `TokenSelectorItem.tsx`, sets `numberOfLines={2}` and
`ellipsizeMode="tail"` on the token name `Text`
> - Adds tests in `TokenSelectorItem.test.tsx` to verify two-line
truncation and tail ellipsis
> 
> <sup>Written by [Cursor
Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit
a790ede641d717e3a7b0f713d44e1ee4b1a85e43. This will update automatically
on new commits. Configure
[here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

GeorgeGkas committed:
chore: Merge branch 'main' into swaps-new-token-picker

bfullam committed:
chore: use network naming from LD for network pills

GeorgeGkas committed:
chore: Merge branch 'main' into swaps-new-token-picker

bfullam committed:
fix: get network name from chainRanking

bfullam committed:
test: update e2e mocks

GeorgeGkas committed:
fix: do not render loader when market price fetch fail

GeorgeGkas committed:
refactor: extract complex logic into a component

GeorgeGkas committed:
Merge branch 'swaps-new-token-picker' into swaps-3694-no-render-skeleton

GeorgeGkas committed:
fix: do not render loader when market price fetch fail (#24558)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

Do not render loader when market price fetch fail

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the
CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing
description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and
accurately)
-->

CHANGELOG entry: do not render loader when market price fetch fail

## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/SWAPS-3694

## **Manual testing steps**

```gherkin
Feature: my feature name

  Scenario: user [verb for user action]
    Given [describe expected initial app state]

    When user [verb for user action]
    Then [describe expected outcome]
```

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.


<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Prevents misleading loaders when market data fails and tightens
balance rendering logic.
> 
> - Extracts `FiatBalanceView` in `TokenSelectorItem` to render a
skeleton only for `TOKEN_BALANCE_LOADING` states and render nothing for
`TOKEN_RATE_UNDEFINED`; integrates it into the balance section
> - Updates `TokenSelectorItem.test.tsx` to cover lowercase/uppercase
loading tokens, undefined rate, available fiat balance, and
`shouldShowBalance` off
> - Adjusts `deriveBalanceFromAssetMarketDetails` to return
`TOKEN_RATE_UNDEFINED` for tokens when `conversionRate` is missing
(preserving ETH behavior); updates tests to match
> 
> <sup>Written by [Cursor
Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit
9061bb5df04881c890c893637af0cead3abbfc7c. This will update automatically
on new commits. Configure
[here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
chore: bump bridge-controller

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
chore: dedupe

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

GeorgeGkas committed:
Merge branch 'main' into swaps-new-token-picker

brianacnguyen commented on app/components/UI/Bridge/components/BridgeTokenSelector/BridgeTokenSelector.tsx:
```diff
@@ -0,0 +1,529 @@
+
+  // Set navigation options for header
+  useEffect(() => {
+    navigation.setOptions(getBridgeTokenSelectorNavbar(navigation));
```
can you update this to use HeaderCenter inline?

brianacnguyen commented on app/components/UI/Bridge/components/BridgeTokenSelector/NetworkPills.tsx:
```diff
@@ -0,0 +1,105 @@
+      const isSelected = selectedChainId === chain.chainId;
+
+      return (
+        <Pressable
```
you can just use ButtonToggle for this

brianacnguyen reviewed (dismissed):
Requesting a few changes RE using DS components

Prithpal-Sooriya reviewed (dismissed):
Asset CO changes LGTM.

bfullam committed:
test: update test

GeorgeGkas reviewed (dismissed):
![Image](https://github.com/user-attachments/assets/8dc17d21-1395-4bd5-9eab-6f1ff682f1e0)

bfullam committed:
chore: design system updates

bfullam commented on app/components/UI/Bridge/components/BridgeTokenSelector/BridgeTokenSelector.tsx:
```diff
@@ -0,0 +1,529 @@
+    }
+
+    return (
+      <Box
```
Done

bfullam committed:
test: fix test

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
Revert "test: fix test"

This reverts commit 755bfdac72d0ef5ce96326be344cd0d5cd8db9c8.

bfullam committed:
test: fix tests

bfullam committed:
test: fix e2e api mocking

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

bfullam committed:
chore: bump bridge controller

bfullam committed:
chore: lint

davibroc reviewed (dismissed):
LGTM

bfullam committed:
Merge branch 'main' into swaps-new-token-picker

infiniteflower commented on app/components/UI/Bridge/components/TokenInputArea/index.tsx:
```diff
@@ -253,18 +252,15 @@ export const TokenInputArea = forwardRef<
-        } as BridgeDestNetworkSelectorRouteParams,
+    const navigateToDestTokenSelector = () => {
+      navigation.navigate(Routes.BRIDGE.TOKEN_SELECTOR, {
+        type: 'dest',
```
Non blocking comment: Can we use enums instead of raw strings?

gambinish reviewed (approved):
Approving to unblock

infiniteflower commented on app/components/UI/Bridge/hooks/useSearchTokens.ts:
```diff
@@ -0,0 +1,201 @@
+        }
+
+        const response = await fetch(
+          `${process.env.BRIDGE_USE_DEV_APIS === 'true' ? BRIDGE_DEV_API_BASE_URL : BRIDGE_PROD_API_BASE_URL}/getTokens/search`,
```
Same as above

infiniteflower commented on app/core/redux/slices/bridge/index.ts:
```diff
@@ -272,6 +281,43 @@ export const selectBridgeFeatureFlags = createSelector(
+        const hexChainId =
+          `0x${parseInt(decimalChainId, 10).toString(16)}` as Hex;
+        return enabledChainIds.has(hexChainId);
+      }
```
Same as above, we can use the util from bridge controller to convert to hex chain id
